### PR TITLE
BlockId removal: refactor: CallExecutor trait

### PR DESCRIPTION
**Pull request by michalkucharczyk**
It changes the arguments of CallExecutor methods:
-  `call`, `contextual_call`, `runtime_version`, `prove_execution`

from: `BlockId<Block>` to: `Block::Hash`

This PR is part of BlockId::Number refactoring analysis (paritytech/substrate#11292)

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2285821

File: client/finality-grandpa/src/lib.rs [hunk 1]
```diff
@@ -464,18 +464,18 @@ pub trait GenesisAuthoritySetProvider<Block: BlockT> {
 	fn get(&self) -> Result<AuthorityList, ClientError>;
 }
 
-impl<Block: BlockT, E> GenesisAuthoritySetProvider<Block>
-	for Arc<dyn ExecutorProvider<Block, Executor = E>>
+impl<Block: BlockT, E, Client> GenesisAuthoritySetProvider<Block> for Arc<Client>
```
Comment (michalkucharczyk) [inline client/finality-grandpa/src/lib.rs:467]: I had some doubts while reworking this.

Used in: https://github.com/paritytech/substrate/blob/911c9ab067d3cf487cfef61906ed2525cb87a5c9/bin/node/cli/src/service.rs#L193-L197,
https://github.com/paritytech/substrate/blob/911c9ab067d3cf487cfef61906ed2525cb87a5c9/client/finality-grandpa/src/lib.rs#L496-L501

File: client/api/src/call_executor.rs [hunk 2]
```diff
@@ -54,7 +54,7 @@ pub trait CallExecutor<B: BlockT>: RuntimeVersionOf {
 	/// No changes are made.
 	fn call(
 		&self,
-		id: &BlockId<B>,
+		at_hash: <B as BlockT>::Hash,
```
Comment (bkchr) [inline client/api/src/call_executor.rs:57]:
```suggestion
		at_hash: B::Hash,
```

File: client/api/src/call_executor.rs [hunk 3]
```diff
@@ -67,7 +67,7 @@ pub trait CallExecutor<B: BlockT>: RuntimeVersionOf {
 	/// of the execution context.
 	fn contextual_call(
 		&self,
-		at: &BlockId<B>,
+		at_hash: <B as BlockT>::Hash,
```
Comment (bkchr) [inline client/api/src/call_executor.rs:70]:
```suggestion
		at_hash: B::Hash,
```

File: client/api/src/call_executor.rs [hunk 4]
```diff
@@ -83,14 +83,17 @@ pub trait CallExecutor<B: BlockT>: RuntimeVersionOf {
 	/// Extract RuntimeVersion of given block
 	///
 	/// No changes are made.
-	fn runtime_version(&self, id: &BlockId<B>) -> Result<RuntimeVersion, sp_blockchain::Error>;
+	fn runtime_version(
+		&self,
+		at_hash: <B as BlockT>::Hash,
```
Comment (bkchr) [inline client/api/src/call_executor.rs:88]:
```suggestion
		at_hash: B::Hash,
```

File: client/api/src/call_executor.rs [hunk 5]
```diff
@@ -83,14 +83,17 @@ pub trait CallExecutor<B: BlockT>: RuntimeVersionOf {
...
+	) -> Result<RuntimeVersion, sp_blockchain::Error>;
 
 	/// Prove the execution of the given `method`.
 	///
 	/// No changes are made.
 	fn prove_execution(
 		&self,
-		at: &BlockId<B>,
+		at_hash: <B as BlockT>::Hash,
```
Comment (bkchr) [inline client/api/src/call_executor.rs:96]:
```suggestion
		at_hash: B::Hash,
```

File: client/service/src/client/call_executor.rs [hunk 6]
```diff
@@ -117,12 +117,11 @@ where
 	/// Returns the on chain runtime version.
 	fn on_chain_runtime_version(
 		&self,
-		id: &BlockId<Block>,
+		hash: <Block as BlockT>::Hash,
```
Comment (bkchr) [inline client/service/src/client/call_executor.rs:120]:
```suggestion
		hash: Block::Hash,
```

File: client/service/src/client/call_executor.rs [hunk 7]
```diff
@@ -166,21 +165,21 @@ where
 
 	fn call(
 		&self,
-		at: &BlockId<Block>,
+		at_hash: <Block as BlockT>::Hash,
```
Comment (bkchr) [inline client/service/src/client/call_executor.rs:168]:
```suggestion
		at_hash: Block::Hash,
```

File: client/service/src/client/call_executor.rs [hunk 8]
```diff
@@ -206,7 +205,7 @@ where
 
 	fn contextual_call(
 		&self,
-		at: &BlockId<Block>,
+		at_hash: <Block as BlockT>::Hash,
```
Comment (bkchr) [inline client/service/src/client/call_executor.rs:208]:
```suggestion
		at_hash: Block::Hash,
```

File: client/service/src/client/call_executor.rs [hunk 9]
```diff
@@ -275,32 +274,34 @@ where
 		.map_err(Into::into)
 	}
 
-	fn runtime_version(&self, id: &BlockId<Block>) -> sp_blockchain::Result<RuntimeVersion> {
-		let at_hash = self.backend.blockchain().expect_block_hash_from_id(id)?;
+	fn runtime_version(
+		&self,
+		at_hash: <Block as BlockT>::Hash,
```
Comment (bkchr) [inline client/service/src/client/call_executor.rs:279]:
```suggestion
		at_hash: Block::Hash,
```

File: client/service/src/client/call_executor.rs [hunk 10]
```diff
@@ -275,32 +274,34 @@ where
...
 		let runtime_code =
 			state_runtime_code.runtime_code().map_err(sp_blockchain::Error::RuntimeCode)?;
-		self.check_override(runtime_code, id).map(|(_, v)| v)
+		self.check_override(runtime_code, at_hash).map(|(_, v)| v)
 	}
 
 	fn prove_execution(
 		&self,
-		at: &BlockId<Block>,
+		at_hash: <Block as BlockT>::Hash,
```
Comment (bkchr) [inline client/service/src/client/call_executor.rs:291]:
```suggestion
		at_hash: Block::Hash,
```

File: client/service/src/client/call_executor.rs [hunk 11]
```diff
@@ -340,7 +341,10 @@ where
 	E: CodeExecutor + RuntimeVersionOf + Clone + 'static,
 	Block: BlockT,
 {
-	fn runtime_version(&self, at: &BlockId<Block>) -> Result<sp_version::RuntimeVersion, String> {
+	fn runtime_version(
+		&self,
+		at: <Block as BlockT>::Hash,
```
Comment (bkchr) [inline client/service/src/client/call_executor.rs:346]:
```suggestion
		at: Block::Hash,
```

**Review by bkchr — APPROVED** [review]

**Review by dmitry-markin — APPROVED** [review]

**Review by melekes — APPROVED** [review]
👍

**Comment by michalkucharczyk** [comment]
bot rebase

**Comment by michalkucharczyk** [comment]
bot merge

**Pull request merged**